### ISSUE TITLE
Accessing the content of a JSON-document

### ISSUE THREAD

**Issue by xml-project** [xproc/3.0-specification#392]
Currently we define four document types, but only the content (representation) of XML and text documents are accessible in XPath expressions (say to set a variable or use it in p:when).

Obviously there are use cases, where I would like to access the representation of a JSON doc in a variable (extracting the value of one key from a map). To do this would currently involve a p:cast-content-type where I convert the JSON document to an XML document and then use XPath on that document.

So I wonder if it would not be useful to have not only a function `p:document-properties()`, but also a function `p:document-representation($doc as document-node()) as xs:item` with the following semantics:

1. if $doc has an XML mediatype or a Text mediatype the document-node is returned.
2. if $doc has a JSON mediatype the function returns the JSON document according to the rules set out by [fn:parse-json](https://www.w3.org/TR/xpath-functions-31/#func-parse-json).
3. for any other media type it is implementation defined how the representation is presented in XDM.

Do we want this as a standard of should we leave it to the implementors, whether and how a the representation of a JSON/binary document is accessed in XPath?

**Comment by gimsieke** [comment]
We’ve been contemplating a `p:parse-json()` function [in London](https://github.com/xproc/Workshop-2017-06/wiki/Minutes).

**Comment by xml-project** [comment]
Thanks for reminding me about this, I forgot totally. We definitely had the wrong name, because p:parse-json and fn:parse-json are quite different, but this is just the question of naming the function.

The example we used in London now found a much better solution (without a JSON document):
````
<p:variable name="var" select="fn:json-doc('some.json')" />
````

We had no decision on this in London and I could not find an issue for this case, so I created one.
Do you think, the special function only for JSON-documents is better?
I am not sure what to do about binary documents. There might be use cases where I would like to access them in XPath too and this is only possible with the

**Comment by xml-project** [comment]
I think it is something completely different. fn:parse-json() takes an instance of xs:string and returns the parsed JSON.
The function a proposed takes an JSON document as part of the XProc model and exposes it to XPath. Like so:
````
<p:load href="some.json" /> <!-- creates an XProc JSON document -->

<p:choose>
  <p:when test="map:get(this-is-the-proposed-function(/), 'key')) = 'some value'">
     ...
  </p:when>
</p:choose>
````
What did I miss?

**Comment by ndw** [comment]
At the 11 June f2f, David argues for a JSON function, not a general function. Norm proposes `p:json()`, Aachim proposes `p:get-json()` on the basis that just "json" isn't clear enough.

Semantics: `p:_json_()` takes a JSON document and returns the XDM item that represents that JSON document. It is an error otherwise.

Consensus is: `p:json()`

**Comment by ndw** [comment]
Recent improvements to `p:cast-content-type` make it possible to cast JSON to XML (and vice-versa) with parameters to control which mapping you'd like to use.

Looking back at the motivating example:

```xml

<p:load href="some.json" /> <!-- creates an XProc JSON document -->

<p:choose>
  <p:when test="map:get(this-is-the-proposed-function(/), 'key')) = 'some value'">
     ...
  </p:when>
</p:choose>
```

It could be rewritten easily to use cast-content-type:

```xml

<p:load href="some.json" /> <!-- creates an XProc JSON document -->

<p:cast-content-type content-type="application/xml"/>

<p:choose>
  <p:when test="/json/key = 'some value'">
     ...
  </p:when>
</p:choose>
```

Despite consensus to add `p:json()`, I think it's unnecessary. Anyone else persuaded?

If not, and we press on with `p:json()`, I think it should have a single, optional `map()` argument so that the caller can select which kind of JSON to XML transformation they'd like.

And if we have `p:json()` to convert JSON to XML, why isn't a corresponding function to get from XML back to JSON needed?

**Comment by xml-project** [comment]
I still think the function the function is necessary because I want to access JSON, not convert JSON to XML to access it. If you do not want it, it will be a processor specific function in my implementation.

**Comment by ndw** [comment]
See #467 
I'm not saying I don't want it, I'm just trying to work out what it's for/how it works

**Comment by xml-project** [comment]
I do not like the function solution either and would like to access the XDM item representing the JSON directly. The decision which got us there is that we need a document node as a parameter for p:document-properties().

I remember us having discussed this two years ago in Prague: One idea was to augment the dynamic context of XPath expressions in XProc, so that the document properties of a document are part of the context and are accessible via document-properties() without any parameter. In this case "." could be any non document instance of the XDM universe, but since it has to be a document node, I think a function "p:json($doc as document-node())) as xs:item" is indispensable.

The item returned is the same as for fn:parse-json(), so its either a map, an array, xs:string, xs:double, xs:boolean, [Oxford comma !!! ;-))] or an empty sequence. I do not think that a JSON-> XML transformation needs to take place.

**Comment by xml-project** [comment]
I think this is fixed with pr #475
